test/libsolidity/util/TestFileParser.cpp

```diff
@@ -0,0 +1,392 @@
+#include <boost/throw_exception.hpp>
+#include <fstream>
+#include <memory>
+#include <stdexcept>
+
+using namespace dev;
+using namespace langutil;
+using namespace solidity;
+using namespace dev::solidity::test;
+using namespace std;
+
+namespace
+{
+	bool isDecimalDigit(char c)
+	{
+		return '0' <= c && c <= '9';
+	}
+	bool isWhiteSpace(char c)
+	{
+		return c == ' ' || c == '\n' || c == '\t' || c == '\r';
+	}
+	bool isIdentifierStart(char c)
+	{
+		return c == '_' || c == '$' || ('a' <= c && c <= 'z') || ('A' <= c && c <= 'Z');
+	}
+	bool isIdentifierPart(char c)
+	{
+		return isIdentifierStart(c) || isDecimalDigit(c);
+	}
+}
+
+vector<dev::solidity::test::FunctionCall> TestFileParser::parseFunctionCalls()
+{
+	vector<FunctionCall> calls;
+	if (!accept(SoltToken::EOS))
+	{
+		// TODO: check initial token state
+		expect(SoltToken::Unknown);
+		while (!accept(SoltToken::EOS))
+		{
+			if (!accept(SoltToken::Whitespace))
+			{
+				FunctionCall call;
+
+				/// If this is not the first call in the test,
+				/// the last call to parseParameter could have eaten the
+				/// new line already. This could only be fixed with a one
+				/// token lookahead that checks parseParameter
+				/// if the next token is an identifier.
+				if (calls.empty())
+					expect(SoltToken::Newline);
+				else
+					accept(SoltToken::Newline, true);
+
+				call.signature = parseFunctionSignature();
+				if (accept(SoltToken::Comma, true))
+					call.value = parseFunctionCallValue();
+				if (accept(SoltToken::Colon, true))
+					call.arguments = parseFunctionCallArguments();
+
+				if (accept(SoltToken::Newline, true))
+					call.displayMode = FunctionCall::DisplayMode::MultiLine;
+
+				call.arguments.comment = parseComment();
+
+				if (accept(SoltToken::Newline, true))
+					call.displayMode = FunctionCall::DisplayMode::MultiLine;
+
+				expect(SoltToken::Arrow);
+				call.expectations = parseFunctionCallExpectations();
+				call.expectations.comment = parseComment();
+
+				calls.emplace_back(std::move(call));
+			}
+			else
+				m_scanner.scanNextToken();
+		}
+	}
+	return calls;
+}
+
+string TestFileParser::formatToken(SoltToken _token)
+{
+	switch (_token)
+	{
+#define T(name, string, precedence) case SoltToken::name: return string;
+		SOLT_TOKEN_LIST(T, T)
+#undef T
+		default: // Token::NUM_TOKENS:
+			return "";
+	}
+}
+
+bool TestFileParser::accept(SoltToken _token, bool const _expect)
+{
+	if (m_scanner.currentToken() == _token)
+	{
+		if (_expect)
+			expect(_token);
+		return true;
+	}
+	return false;
+}
+
+bool TestFileParser::expect(SoltToken _token, bool const _advance)
+{
+	if (m_scanner.currentToken() != _token)
+		throw Error(
+			Error::Type::ParserError,
+			"Unexpected " + formatToken(m_scanner.currentToken()) + ": \"" +
+			m_scanner.currentLiteral() + "\". " +
+			"Expected \"" + formatToken(_token) + "\"."
+			);
+	if (_advance)
+		m_scanner.scanNextToken();
+	return true;
+}
+
+string TestFileParser::parseFunctionSignature()
```

Comment:
Actually, I think this should maybe not try to parse the actual signature at all, but just read the whole string in the parenthesis and directly store it as signature. That way we can already call a function with signature ``f(uint8)``, although the value for the argument can only be given as ``uint256`` in the end (but arguments are padded anyways, so no problem there).

Reply:
I do also think that it would be nice to be able to call `f(uint8)`. On the other hand, this would almost come for free with a new token for `uint8`. Can be handled as `u256` interally. Should we include that?